### DISCUSSION TITLE
DeviceClient.ReceiveAsync ignores token cancellation

### DISCUSSION THREAD

**Discussion by dkattan** [Azure/azure-iot-sdk-csharp#2108] [General]
# Context

- .NET 5
- DeviceClient SDK 1.36
- AMQP

## Description of the issue
DeviceClient.ReceiveAsync(CancellationToken cancellationToken) method does not exit when cancellationToken is cancelled.

The current tests simply verify that the task returns immediately when passed a cancellationToken that is already cancelled.

## Code sample exhibiting the issue

```
protected override async Task ExecuteAsync(CancellationToken stoppingToken)
{
      stoppingToken.Register(() => { _logger.LogInformation("Stopping Token Called"); });
      while (!stoppingToken.IsCancellationRequested)
      {
        using Message receivedMessage = await DeviceClient.ReceiveAsync(stoppingToken);
        if (receivedMessage == null) continue;
      }
     _logger.LogInformation(“Exiting Application”);
}
```

## Console log of the issue
```
Stopping Token Called
```

## Expected Result
```
Stopping Token Called
Exiting Application
```

## Workaround
To workaround this I have changed my code to use the ReceiveAsync(TimeSpan) overload, but it feels hacky.

**Comment by timtay-microsoft** [comment]
Looks like the SDK uses the default receive timeout set in the transport settings rather than the cancellation token when using the receiveAsync API.

https://github.com/Azure/azure-iot-sdk-csharp/blob/118a15e4e46a0cd3a6e6800c430eb234c232e274/iothub/device/src/Transport/Amqp/AmqpTransportHandler.cs#L206

That seems like a bug to me as well, so I'll fix this

**Comment by timtay-microsoft** [comment]
After further investigation, the real issue here appears to be that the AMQP library we use simply doesn't accept cancellation tokens. Instead, it only accepts TimeSpans as you can see [here](https://github.com/Azure/azure-amqp/blob/ee0de3f71b930ec6c78bd9f4b3120ac873b50725/src/ReceivingAmqpLink.cs#L165). I'll discuss this a bit more with my team.

**Comment by timtay-microsoft** [comment]
Looks like there is already an ask out for adding support for cancellation tokens to the AMQP library in question [here](https://github.com/Azure/azure-amqp/issues/181). Ideally they would add support and we could just pass along the cancellation token as expected.
